### PR TITLE
Add Datum Arithmetic tests, Fix Interval Substraction (#4480)

### LINKED ISSUE #1065: Incorrect Numeric operations on MONTH_DAY_NANO interval type
**Describe the bug**
Pointed out by @jorgecarleitao  and @b41sh  on https://github.com/apache/arrow-rs/pull/779#discussion_r763641073

> Semantically, the numerics of an i128 are not the same as the numerics of (months,days,nanos) since i128 + i128 != (months, days,nanos) + (months, days,nanos).

> The consequence of defining this type numerically here is that arithmetic kernels will accept this type, but they will yield a semantically incorrect result (e.g. i128 + i128 to sum two intervals of 1 month each).

**To Reproduce**
Do a numerical operation on two arrays of `MonthDayNano` type (for example add or subtract them). The arithmetic will be performed directly on the `i128` representation rather than field by field

**Expected behavior**
The intervals should be calculated field by field; So for example to add two `MonthDayNano` fields, the kernels should add the nanoseconds, days and months separately

**Additional context**

### LINKED ISSUE #4480: Port Tests from Deprecated Arithmetic Kernels
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

https://github.com/apache/arrow-rs/pull/4465 deprecated the old arithmetic kernels, but continued to use them to provide test coverage of the new APIs.

**Describe the solution you'd like**

We should port the old tests / add new tests to allow the eventual removal of the now deprecated arithmetic kernels

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #4489: Incorrect Interval Subtraction
**Describe the bug**

There are a couple of bugs in the interval subtraction logic:

* It assumes negating the value negates its components, see #4488 for why this is incorrect
* It assumes negation cannot overflow
* TimestampSecondType::subtract_day_time uses microseconds instead of milliseconds

**To Reproduce**

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4480 
Closes #4489
Closes #1065

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-arith/src/arithmetic.rs [hunk 1]
```diff
@@ -1315,7 +1315,7 @@ mod tests {
         let c = c.as_any().downcast_ref::<Date32Array>().unwrap();
         assert_eq!(
             c.value(0),
-            Date32Type::from_naive_date(NaiveDate::from_ymd_opt(2023, 3, 27).unwrap())
+            Date32Type::from_naive_date(NaiveDate::from_ymd_opt(2023, 3, 28).unwrap())
```
Comment (tustvold) [inline arrow-arith/src/arithmetic.rs:1318]: The previous logic was incorrect but happened to give a potentially plausible result. I'm not sure what the correct rounding behaviour is here though

File: arrow-arith/src/numeric.rs [hunk 2]
```diff
@@ -89,6 +90,18 @@ enum Op {
     Rem,
 }
 
+impl std::fmt::Display for Op {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
```
Comment (alamb) [inline arrow-arith/src/numeric.rs:94]: 💯

File: arrow-arith/src/numeric.rs [hunk 3]
```diff
@@ -570,8 +582,8 @@ fn decimal_op<T: DecimalType>(
                     .saturating_add(1)
                     .min(T::MAX_PRECISION);
 
-            let l_mul = T::Native::usize_as(10).pow_wrapping((result_scale - s1) as _);
-            let r_mul = T::Native::usize_as(10).pow_wrapping((result_scale - s2) as _);
+            let l_mul = T::Native::usize_as(10).pow_checked((result_scale - s1) as _)?;
```
Comment (alamb) [inline arrow-arith/src/numeric.rs:585]: I was a little confused at first about why  both `AddWrapping` and `Add` using the same function. Then 
I realized this is for computing the scale factor

Given `MAX_PRECISION` is a constant, what is the reason to change this to `pow_checked`? Maybe it doesn't matter performance wise, I was just curious

File: arrow-arith/src/numeric.rs [hunk 4]
```diff
@@ -670,3 +682,490 @@ fn decimal_op<T: DecimalType>(
...
+
+        let a = Int8Array::from(vec![Some(2), None, Some(45)]);
+        let b = Int8Array::from(vec![Some(5), Some(3), None]);
+        let result = add(&a, &b).unwrap();
+        assert_eq!(result.as_ref(), &Int8Array::from(vec![Some(7), None, None]));
+
+        let a = UInt8Array::from(vec![56, 5, 3]);
+        let b = UInt8Array::from(vec![200, 2, 5]);
+        let err = add(&a, &b).unwrap_err().to_string();
+        assert_eq!(err, "Compute error: Overflow happened on: 56 + 200");
```
Comment (alamb) [inline arrow-arith/src/numeric.rs:718]: 👨‍🍳 👌  very nice message

File: arrow-arith/src/numeric.rs [hunk 5]
```diff
@@ -670,3 +682,490 @@ fn decimal_op<T: DecimalType>(
...
+
+        let a = Int16Array::from(vec![i16::MIN]);
+        let b = Int16Array::from(vec![-1]);
+        let err = div(&a, &b).unwrap_err().to_string();
+        assert_eq!(err, "Compute error: Overflow happened on: -32768 / -1");
+
+        let a = Int16Array::from(vec![21]);
+        let b = Int16Array::from(vec![0]);
+        let err = div(&a, &b).unwrap_err().to_string();
+        assert_eq!(err, "Divide by zero error");
```
Comment (alamb) [inline arrow-arith/src/numeric.rs:744]: I didn't see a test for integer remainder `%`

File: arrow-arith/src/numeric.rs [hunk 6]
```diff
@@ -670,3 +682,490 @@ fn decimal_op<T: DecimalType>(
...
+        );
+
+        let a = Decimal128Array::from(vec![1])
+            .with_precision_and_scale(3, 3)
+            .unwrap();
+        let b = Decimal128Array::from(vec![1])
+            .with_precision_and_scale(37, 37)
+            .unwrap();
+        let err = mul(&a, &b).unwrap_err().to_string();
+        assert_eq!(err, "Invalid argument error: Output scale of Decimal128(3, 3) * Decimal128(37, 37) would exceed max scale of 38");
```
Comment (alamb) [inline arrow-arith/src/numeric.rs:837]: 👍

File: arrow-arith/src/numeric.rs [hunk 7]
```diff
@@ -670,3 +682,490 @@ fn decimal_op<T: DecimalType>(
...
+        .into_iter()
+        .map(|x| {
+            T::make_value(DateTime::parse_from_rfc3339(x).unwrap().naive_utc()).unwrap()
+        })
+        .collect();
+
+        let a = PrimitiveArray::<T>::new(values, None);
+        let b = IntervalYearMonthArray::from(vec![
+            IntervalYearMonthType::make_value(5, 34),
+            IntervalYearMonthType::make_value(-2, 4),
```
Comment (alamb) [inline arrow-arith/src/numeric.rs:898]: Can we also add a test for overflow (like subtracting two months from `1970-01-01T00:00:00Z`?

File: arrow-arith/src/numeric.rs [hunk 8]
```diff
@@ -670,3 +682,490 @@ fn decimal_op<T: DecimalType>(
...
+        let result = sub(&a, &b).unwrap();
+        assert_eq!(result.as_primitive::<T>().values(), &[996, 4399, -3701]);
+
+        let a = PrimitiveArray::<T>::new(vec![i64::MAX].into(), None);
+        let b = PrimitiveArray::<T>::new(vec![1].into(), None);
+        let err = add(&a, &b).unwrap_err().to_string();
+        assert_eq!(
+            err,
+            "Compute error: Overflow happened on: 9223372036854775807 + 1"
+        );
```
Comment (alamb) [inline arrow-arith/src/numeric.rs:1066]: is it worthing adding a test here showing that `*` and `/` are not supported for durations?

File: arrow-array/src/types.rs [hunk 9]
```diff
@@ -457,17 +457,17 @@ impl TimestampSecondType {
...
         })?;
         let res = res
-            .checked_add_signed(Duration::days(days as i64))
+            .checked_sub_signed(Duration::days(days as i64))
             .ok_or_else(|| {
                 ArrowError::ComputeError("Timestamp out of range".to_string())
             })?;
         let res = res
-            .checked_add_signed(Duration::microseconds(ms as i64))
+            .checked_sub_signed(Duration::milliseconds(ms as i64))
```
Comment (alamb) [inline arrow-array/src/types.rs:470]: This is a bug fix -- that `IntervalDayTime` stores data in milliseconds not microseconds 👍

**Review by alamb — APPROVED** [review]
Thank you @tustvold  -- this looks really nice to me

File: arrow-arith/src/numeric.rs (hunk 7 above)
Comment (tustvold) [reply arrow-arith/src/numeric.rs:898]: Currently the overflow behaviour is a little broken, #4456 tracks fixing this.

File: arrow-arith/src/numeric.rs (hunk 3 above)
Comment (tustvold) [reply arrow-arith/src/numeric.rs:585]: It can overflow on negative scales

**Pull request merged**